File: my_cpp_practice/src/add_two_ints_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"

void delay_ms(unsigned long milliseconds)
{
  usleep(milliseconds * 1000);
}

class AddTwoIntsClientNode : public rclcpp::Node
{
public:
  AddTwoIntsClientNode() : Node("add_two_ints_client")
  {
    // thread1_ = std::thread(std::bind(&AddTwoIntsClientNode::call_add_two_ints_service, this, 1, 4));
    threads_.push_back(std::thread(std::bind(&AddTwoIntsClientNode::call_add_two_ints_service, this, 1, 4)));
    threads_.push_back(std::thread(std::bind(&AddTwoIntsClientNode::call_add_two_ints_service, this, 4, 5)));
  }

  void call_add_two_ints_service(int req_a, int req_b)
  {
    auto client = this->create_client<example_interfaces::srv::AddTwoInts>("add_two_ints");
    while (!client->wait_for_service(std::chrono::seconds(1)))
    {
      RCLCPP_WARN(this->get_logger(), "Waiting for the add_two_int_server to be up...");
    }

    auto req = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
    req->a = req_a;
    req->b = req_b;

    auto future = client->async_send_request(req);

    try
    {
      auto res = future.get();
      RCLCPP_INFO(this->get_logger(), "%d + %d = %d", (int)req->a, (int)req->b, (int)res->sum);
    }
    catch (const std::exception &e)
    {
      RCLCPP_ERROR(this->get_logger(), "Error while calling add_two_ints_service");
    }
  }

private:
  std::vector<std::thread> threads_;
  // std::thread thread1_;
};

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<AddTwoIntsClientNode>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

File: my_py_practice/my_py_practice/add_two_ints_client.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from functools import partial

from example_interfaces.srv import AddTwoInts


class AddTwoIntsClientNode(Node):
    def __init__(self):
        super().__init__("add_two_ints_client")
        self.call_add_two_ints_server(6, 7)
        self.call_add_two_ints_server(10, 9)
        self.call_add_two_ints_server(67, 13)

    def call_add_two_ints_server(self, req_a, req_b):
        client = self.create_client(AddTwoInts, "add_two_ints")
        while not client.wait_for_service(1.0):
            self.get_logger().warn("waithing for add_two_int_server ...")

        req = AddTwoInts.Request()
        req.a = req_a
        req.b = req_b

        future = client.call_async(req)
        future.add_done_callback(
            partial(self.callback_call_add_two_ints, req_a=req.a, req_b=req.b))

    def callback_call_add_two_ints(self, future, req_a, req_b):
        try:
            res = future.result()
            self.get_logger().info("%ld + %ld = %ld" % (req_a, req_b, res.sum))
        except Exception as e:
            self.get_logger().error("Service call failed %r" % (e,))


def main(args=None):
    rclpy.init(args=args)
    node = AddTwoIntsClientNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()